### PR TITLE
8345573: Module dependencies not resolved from run-time image when --limit-module is being used

### PR DESCRIPTION
**Pull request by jerboaa**
Please review this fix for JEP 493 enabled JDKs related to the `--limit-modules` option. The existing jlink `bindservices` tests cover this issue. Previously they didn't run on a JEP 493 enabled JDK, since `jmods` folder is missing for them.

The gist of the issue is that multiple `ModuleFinder`s were at play. In particular, when linking from the run-time image a finder is being set up based solely on the module-path. Which in that case only contains the application modules (not the JDK dependencies). So if the `--limit-modules` clause included any JDK modules that the app needs as a dependency the finder would fail because the module specified in the `--limit-modules` clause didn't exist in the finder's look-up space.

[A similar case happens](https://bugs.openjdk.org/browse/JDK-8345573?focusedId=14729247&page=com.atlassian.jira.plugin.system.issuetabpanels:comment-tabpanel#comment-14729247) for a regular JDK build where `ALL-MODULE-PATH` is being used together with `--limit-modules`. In that case, the `FindException` is being thrown when the `root` set of modules are attempted to get assembled.

The fix that I'm proposing here is two fold:

1. Distinguish whether or not we ought to resolve the JDK modules from the run-time image or not.
2. Based on that information set up the finders as appropriate and use the same finders when assembling the module root set.

I think this makes for a cleaner implementation and also fixes the `--limit-modules` issue when linking from the run-time image as any finder that is being limited has already been properly set up to include dependencies (if need be).

Testing:

- [x]  GHA
- [x] jlink tests in `test/jdk/tools/jlink` on a regular JDK build and on a build with `--enable-linkable-runtime`. Both test runs show all tests passing.

Thoughts?

---------
### Progress
- [x] Change must be properly reviewed (1 review required, with at least 1 [Reviewer](https://openjdk.org/bylaws#reviewer))
- [x] Change must not contain extraneous whitespace
- [x] Commit message must refer to an issue

### Issue
 * [JDK-8345573](https://bugs.openjdk.org/browse/JDK-8345573): Module dependencies not resolved from run-time image when --limit-module is being used (**Bug** - P3)

### Reviewers
 * [Mandy Chung](https://openjdk.org/census#mchung) (@mlchung - **Reviewer**)

### Reviewing
<details><summary>Using <code>git</code></summary>

Checkout this PR locally: \
`$ git fetch https://git.openjdk.org/jdk.git pull/22609/head:pull/22609` \
`$ git checkout pull/22609`

Update a local copy of the PR: \
`$ git checkout pull/22609` \
`$ git pull https://git.openjdk.org/jdk.git pull/22609/head`

</details>
<details><summary>Using Skara CLI tools</summary>

Checkout this PR locally: \
`$ git pr checkout 22609`

View PR using the GUI difftool: \
`$ git pr show -t 22609`

</details>
<details><summary>Using diff file</summary>

Download this PR as a diff file: \
<a href="https://git.openjdk.org/jdk/pull/22609.diff">https://git.openjdk.org/jdk/pull/22609.diff</a>

</details>
<details><summary>Using Webrev</summary>

[Link to Webrev Comment](https://git.openjdk.org/jdk/pull/22609#issuecomment-2523543838)
</details>

### CODE REVIEW

**Comment by jerboaa** [comment]
GHA failed with:

```
Unknown test selection: 'test/hotspot/jtreg/:tier1_compiler_not_xcomp'
```

Seems unrelated. I'll merge master later today/tomorrow which might fix this.

**Comment by jerboaa** [comment]
> I'll merge master later today/tomorrow which might fix this.

And it did indeed fix it.

**Comment by jerboaa** [comment]
@mlchung Does this look OK to you? Thanks!

File: test/jdk/tools/jlink/bindservices/SuggestProviders.java [hunk 1]
```diff
@@ -54,18 +58,23 @@ public class SuggestProviders {
...
-    private static boolean hasJmods() {
-        if (!Files.exists(Paths.get(JAVA_HOME, "jmods"))) {
-            System.err.println("Test skipped. NO jmods directory");
-            return false;
+    private static boolean isApplicable() {
+        if (!JMODS_EXIST) {
+            if (!LINKABLE_RUNTIME) {
+                System.err.println("Test skipped. Not a linkable runtime and no JMODs");
+                return false;
+            }
```
Comment (mlchung) [inline test/jdk/tools/jlink/bindservices/SuggestProviders.java:77]:
```suggestion
    private static boolean isExplodedJDKImage() {
        if (!JMODS_EXIST && !LINKABLE_RUNTIME) {
            System.err.println("Test skipped. Not a linkable runtime and no JMODs");
            return true;
        }
        return false;
```

This method name tells the conditions when to skip the test.

File: test/jdk/tools/jlink/runtimeImage/AbstractLinkableRuntimeTest.java [hunk 2]
```diff
@@ -192,7 +192,8 @@ protected Path jlinkUsingImage(JlinkSpec spec, OutputAnalyzerHandler handler, Pr
             // if the exit checker failed, we expected the other outcome
             // i.e. fail for success and success for fail.
             boolean successExit = analyzer.getExitValue() == 0;
-            String msg = String.format("Expected jlink to %s given a jmodless image. Exit code was: %d",
+            String msg = String.format("Expected jlink to %s given a run-time image " +
+                                       "link capable image. Exit code was: %d",
```
Comment (mlchung) [inline test/jdk/tools/jlink/runtimeImage/AbstractLinkableRuntimeTest.java:196]:
```suggestion
            String msg = String.format("Expected jlink to %s given a linkable run-time image. " +
                                       "Exit code was: %d",
```

File: test/jdk/tools/jlink/bindservices/BindServices.java [hunk 3]
```diff
@@ -56,19 +58,23 @@ public class BindServices {
...
-    private static boolean hasJmods() {
-        if (!Files.exists(Paths.get(JAVA_HOME, "jmods"))) {
-            System.err.println("Test skipped. NO jmods directory");
-            return false;
+    private static boolean isApplicable() {
+        if (!JMODS_EXIST) {
+            if (!LINKABLE_RUNTIME) {
+                System.err.println("Test skipped. Not a linkable runtime and no JMODs");
+                return false;
+            }
```
Comment (mlchung) [inline test/jdk/tools/jlink/bindservices/BindServices.java:77]:
```suggestion
    private static boolean isExplodedJDKImage() {
        if (!JMODS_EXIST && !LINKABLE_RUNTIME) {
            System.err.println("Test skipped. Not a linkable runtime and no JMODs");
            return true;
        }
        return false;
```

This method name tells the conditions when to skip the test.

File: src/jdk.jlink/share/classes/jdk/tools/jlink/internal/JlinkTask.java [hunk 4]
```diff
@@ -510,23 +483,23 @@ public static Path getDefaultModulePath() {
...
+     * specified in {@code limitMods} plus other modules specified in the
+     * {@code roots} set.
      *
      * @throws IllegalArgumentException if java.base module is present
-     * but its descriptor has no version
+     * but its descriptor has no version or the java.base version is not the
+     * same as the current runtime.
      */
-    public static ModuleFinder newModuleFinder(List<Path> paths,
+    public static ModuleFinder newModuleFinder(ModuleFinder original,
```
Comment (mlchung) [inline src/jdk.jlink/share/classes/jdk/tools/jlink/internal/JlinkTask.java:495]:
```suggestion
    public static ModuleFinder newModuleFinder(ModuleFinder finder,
```

File: src/jdk.jlink/share/classes/jdk/tools/jlink/internal/JlinkTask.java [hunk 5]
```diff
@@ -510,23 +483,23 @@ public static Path getDefaultModulePath() {
...
                                                Set<String> limitMods,
-                                               Set<String> roots)
+                                               Set<String> roots,
+                                               boolean isRuntimeLink)
```
Comment (mlchung) [inline src/jdk.jlink/share/classes/jdk/tools/jlink/internal/JlinkTask.java:498]: `isRuntimeLink` is strictly for performance but this method has no relationship with linking from packaged modules or run-time image.   Suggest to drop this new parameter and refactor the version check as a separate method and call it before creating `JlinkConfiguration`

File: src/jdk.jlink/share/classes/jdk/tools/jlink/internal/JlinkTask.java [hunk 6]
```diff
@@ -429,48 +437,13 @@ private JlinkConfiguration initJlinkConfig() throws BadArgs {
...
-                all.addAll(finder.findAll());
-                return Collections.unmodifiableSet(all);
-            }
-        };
-        // if limitmods is specified then limit the universe
-        if (limitMods != null && !limitMods.isEmpty()) {
-            return limitFinder(combined, limitMods, Objects.requireNonNull(roots));
-        }
-        return combined;
+    public static ModuleFinder createFinderFromPath(List<Path> paths) {
```
Comment (mlchung) [inline src/jdk.jlink/share/classes/jdk/tools/jlink/internal/JlinkTask.java:443]:
```suggestion
    public static ModuleFinder newModuleFinder(List<Path> paths) {
```

**Review by mlchung — COMMENTED** [review]
Thanks for the update.   A couple more comments...

File: src/jdk.jlink/share/classes/jdk/tools/jlink/internal/JlinkTask.java [hunk 7]
```diff
@@ -378,42 +377,51 @@ public static void createImage(JlinkConfiguration config,
...
+            //    modules have not been provided on the module path.
+            // 2. We have a run-time image based link.
+            //
+            // Distinguish case 2 by adding the default 'jmods' folder and try
+            // the look-up again. For case 1 this will now find java.base, but
+            // not for case 2, since the jmods folder is not there or doesn't
+            // include the java.base module.
             Path defModPath = getDefaultModulePath();
             if (defModPath != null) {
                 options.modulePath.add(defModPath);
```
[the remark this answers is not in the document]
Comment (jerboaa) [reply src/jdk.jlink/share/classes/jdk/tools/jlink/internal/JlinkTask.java:396]: Yes, thanks.

File: src/jdk.jlink/share/classes/jdk/tools/jlink/internal/JlinkTask.java (hunk 5 above)
Comment (jerboaa) [reply src/jdk.jlink/share/classes/jdk/tools/jlink/internal/JlinkTask.java:498]: Thanks. I've moved the version check outside this method.

**Comment by jerboaa** [comment]
Thanks for the review, btw.

**Review by mlchung — APPROVED** [review]
Looks good.  Minor renaming suggestion.

File: src/jdk.jlink/share/classes/jdk/tools/jlink/internal/JlinkTask.java [hunk 8]
```diff
@@ -510,48 +492,81 @@ public static Path getDefaultModulePath() {
...
-     * Returns a module finder of the given module path or the system modules
-     * if the module path is empty that limits the observable modules to those
-     * in the transitive closure of the modules specified in {@code limitMods}
-     * plus other modules specified in the {@code roots} set.
+     * Returns a module finder of the given module finder that limits the
+     * observable modules to those in the transitive closure of the modules
+     * specified in {@code limitMods} plus other modules specified in the
+     * {@code roots} set.
+     */
+    public static ModuleFinder limitFinder(ModuleFinder finder,
```
[the remark this answers is not in the document]
Comment (jerboaa) [reply src/jdk.jlink/share/classes/jdk/tools/jlink/internal/JlinkTask.java:500]: Personally, I find it less clear to call it `newModuleFinder`. What it does is that it creates a finder based on the passed in finder that "finds" fewer modules. How about `newLimitedFinder()`? But no strong feelings, so if you think it should be `newModuleFinder` then it's fine by me.

File: src/jdk.jlink/share/classes/jdk/tools/jlink/internal/JlinkTask.java [hunk 9]
```diff
@@ -510,48 +492,81 @@ public static Path getDefaultModulePath() {
...
-    {
-        Runtime.Version version = Runtime.version();
-        Path[] entries = paths.toArray(new Path[0]);
-        ModuleFinder finder = paths.isEmpty() ? ModuleFinder.ofSystem()
-                                              : ModulePath.of(version, true, entries);
-        if (finder.find("java.base").isPresent()) {
-            // use the version of java.base module, if present, as
-            // the release version for multi-release JAR files
-            ModuleDescriptor.Version v = finder.find("java.base").get()
+    private static void checkVersion(ModuleFinder finder) {
```
[the remark this answers is not in the document]
Comment (jerboaa) [reply src/jdk.jlink/share/classes/jdk/tools/jlink/internal/JlinkTask.java:551]: Sure.

File: src/jdk.jlink/share/classes/jdk/tools/jlink/internal/JlinkTask.java (hunk 8 above)
[the remark this answers is not in the document]
Comment (jerboaa) [reply src/jdk.jlink/share/classes/jdk/tools/jlink/internal/JlinkTask.java:500]: I went with `newLimitedFinder`. Hope that's ok.

File: test/jdk/tools/jlink/IntegrationTest.java [hunk 10]
```diff
@@ -154,9 +154,10 @@ private static void test() throws Exception {
         mods.add("java.management");
         Set<String> limits = new HashSet<>();
         limits.add("java.management");
+        boolean linkFromRuntime = false;
         JlinkConfiguration config = new Jlink.JlinkConfiguration(output,
                 mods,
-                JlinkTask.newModuleFinder(modulePaths, limits, mods), false, false, false);
+                JlinkTask.limitFinder(JlinkTask.newModuleFinder(modulePaths), limits, mods), linkFromRuntime, false, false);
```
[the remark this answers is not in the document]
Comment (jerboaa) [reply test/jdk/tools/jlink/IntegrationTest.java]: I've re-formated the code to make this clearer.

**Review by mlchung — APPROVED** [review]
Approve.

Nit: formatting needs some adjustment after the rename.   You can clean up in #22494 and integrate with this version.

File: src/jdk.jlink/share/classes/jdk/tools/jlink/internal/JlinkTask.java (hunk 8 above)
[the remark this answers is not in the document]
Comment (mlchung) [reply src/jdk.jlink/share/classes/jdk/tools/jlink/internal/JlinkTask.java]: This suggestion was mainly caused as this is used by the test which raises the question what is a limit finder. `newModuleFinder` was the original method name and the javadoc specifies what it does.   

Anyway, you can reconsider in #22494.

File: test/jdk/tools/jlink/IntegrationTest.java (hunk 10 above)
[the remark this answers is not in the document]
Comment (mlchung) [reply test/jdk/tools/jlink/IntegrationTest.java]: I see.  Thanks.

**Comment by jerboaa** [comment]
Thank you! I'll do any needed cleanup in #22494

File: src/jdk.jlink/share/classes/jdk/tools/jlink/internal/JlinkTask.java (hunk 8 above)
[the remark this answers is not in the document]
Comment (jerboaa) [reply src/jdk.jlink/share/classes/jdk/tools/jlink/internal/JlinkTask.java]: Thanks.

**Pull request closed without merging**